File: MD_sample/Python_imperative/force.py
import math
from md import Npt, box, rcut2

def dnint(x):
    if (x>0.0):
        return math.floor(x+0.5)
    else:
        return math.ceil(x-0.5)

def force(q):
    dx = [0., 0., 0.]
    for i in range (Npt):
        for k in range (3):
            q[i].ff[k] = 0.

    for i in range (Npt-1):
        for j in range (i+1, Npt):
            r2 = 0.0
            for k in range(3):
                x = q[i].xx[k] - q[j].xx[k]
                dx[k] = x - box * dnint(x/box)
                r2 = r2 + dx[k]*dx[k]
            if (r2 < rcut2):
                coef = 1./(r2*math.sqrt(r2))
                for k in range(3):
                    q[i].ff[k] = q[i].ff[k] + coef * dx[k]
                    q[j].ff[k] = q[j].ff[k] - coef * dx[k]

File: MD_sample/cpp_cmake/src/force.cpp
#include <cmath>
#include "md.h"

void MD::force()
{
  double r2, dxx, dxy, dxz, coef;

  // force initialization 
  for(std::size_t i=0; i<Npt; i++) {
      q.fx[i] = 0.0;
      q.fy[i] = 0.0;
      q.fz[i] = 0.0;
  }

  // coupled loop for many-body interaction
  for (std::size_t i=0; i<Npt-1; i++) {
      for (std::size_t j=i+1; j<Npt; j++) {
	  dxx = q.xx[i] - q.xx[j];
	  dxx -= box * dnint(dxx/box);
	  dxy = q.xy[i] - q.xy[j];
	  dxy -= box * dnint(dxy/box);
	  dxz = q.xz[i] - q.xz[j];
	  dxz -= box * dnint(dxz/box);
	  r2 = dxx*dxx + dxy*dxy + dxz*dxz;
	  if (r2 < rcut2) {
	      coef = 1./(r2*sqrt(r2));
	      q.fx[i] += coef * dxx;
	      q.fy[i] += coef * dxy;
	      q.fz[i] += coef * dxz;
	      q.fx[j] -= coef * dxx;
	      q.fy[j] -= coef * dxy;
	      q.fz[j] -= coef * dxz;
	  }
      }
  }
}

